dashboard/modules/serve/sdk.py

```diff
@@ -0,0 +1,126 @@
+        metadata: Optional[Dict[str, Any]] = None,
+        headers: Optional[Dict[str, Any]] = None,
+    ):
+        if requests is None:
+            raise RuntimeError(
+                "The Serve CLI requires the ray[default] "
+                "installation: `pip install 'ray[default']``"
+            )
+        super().__init__(
+            address=address,
+            create_cluster_if_needed=create_cluster_if_needed,
+            cookies=cookies,
+            metadata=metadata,
+            headers=headers,
+        )
+        self._check_connection_and_version(
+            min_version="1.12",
+            version_error_message="Serve CLI is not supported on the Ray "
+            "cluster. Please ensure the cluster is "
+            "running Ray 1.12 or higher.",
+        )
+
+    def deploy_application(self, app_config: Dict) -> None:
+        deploy_address = f"{self._address}/api/serve/deployments/"
+        response = requests.put(deploy_address, json=app_config)
+
+        if response.status_code == 200:
+            cli_logger.newline()
+            cli_logger.success(
+                "\nSent deploy request successfully!\n "
+                "* Use `serve status` to check your deployments' statuses.\n "
+                "* Use `serve info` to see your running Serve "
+                "application's configuration.\n"
+            )
+            cli_logger.newline()
+        else:
+            self._log_failed_request(response)
+
+    def get_info(self) -> Union[Dict, None]:
+        info_address = f"{self._address}/api/serve/deployments/"
+        response = requests.get(info_address)
+        if response.status_code == 200:
+            return response.json()
+        else:
+            self._log_failed_request(response)
+
+    def get_status(self) -> Union[Dict, None]:
+        status_address = f"{self._address}/api/serve/deployments/status"
+        response = requests.get(status_address)
+        if response.status_code == 200:
+            return response.json()
+        else:
+            self._log_failed_request(response)
+
+    def delete_application(self) -> None:
+        delete_address = f"{self._address}/api/serve/deployments/"
+        response = requests.delete(delete_address)
+        if response.status_code == 200:
+            cli_logger.newline()
+            cli_logger.success("\nSent delete request successfully!\n")
+            cli_logger.newline()
+        else:
+            self._log_failed_request(response)
+
+    def set_up_runtime_env(
+        self,
+        deployment: Deployment,
+        new_runtime_env: Optional[dict] = None,
+        new_working_dir: Optional[str] = None,
+    ) -> None:
+        """
+        1. If not None, new_runtime_env overwrites deployment's runtime_env.
+        2. Then, if not None, new_working_dir overwrites deployment's working_dir.
+        3. Then, uploads the deployment's working_dir if it's local.
+
+        Mutates the deployment.
```

Comment:
This ordering seems confusing to me, IMO it'd be more natural to do:

1. Upload local working_dir if needed.
2. Substitute local working_dir into the passed working_dir.
3. Merge resulting working_dir into deployment one.

We should also only have to do the uploading once per config file, not run it for each individual deployment.

Reply:
By the way, I think we should only override the fields specified in the runtime_env passed to the CLI option. For example, if I have `pip` requirements in the file but no `working_dir`, we should only insert the working dir and leave the `pip`. This enables using `working_dir` for dev in conjunction with other options.

@simon-mo @architkulkarni any thoughts on this?

Reply:
@edoakes re: ordering: Since `_upload_working_dir_if_needed` updates the runtime_env after uploading, it feels strange to manually merge the runtime_env. Are you saying upload the `working_dir` first, and then merge it into all the deployments?  

Although `_upload_working_dir_if_needed` is called once per deployment, the actual upload should only happen once. Do we want to avoid calling `_upload_working_dir_if_needed` more than once?

Reply:
Yes, I think we should still avoid calling `upload_working_dir_if_needed` more than once if we can (it does some amount of caching but may still be slower). Can we modify the passed runtime_env (combination of runtime_env_json and working_dir) before applying it to the deployments? Shouldn't require manual merging, just reordering the calls unless I'm missing something.

Reply:
> By the way, I think we should only override the fields specified in the runtime_env passed to the CLI option. For example, if I have `pip` requirements in the file but no `working_dir`, we should only insert the working dir and leave the `pip`. This enables using `working_dir` for dev in conjunction with other options.
> 
> @simon-mo @architkulkarni any thoughts on this?

I agree this makes sense.

Reply:
I updated `serve run` to call `upload_working_dir_if_needed` only once. I also reordered the runtime_env processing, so it should match the proposal.